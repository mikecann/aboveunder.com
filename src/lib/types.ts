export interface IPost
{
    id: string,
    title: string,
    headerImage: string;
    tags: string[],
    category: string;
    summary: string,
    markdownFileName: string;
    dateCreated: string,
}

export interface IPrint
{
    id: string,
    image:string;
    thumb:string;
    title: string,
    featured: boolean,
    description: string,
    dateCreated: string,
    printOptions: IPrintOption[]
}

export interface IPrintOption
{
    id:string,
    name: string;
    sizes: IPrintOptionSize[]
}

export interface IPrintOptionSize
{
    id:string,
    widthInches: number,
    heightInches: number,
    priceAUD: number,
    weight: number;
}

export interface IDB
{
    posts: IPost[],
    products: IPrint[]
}