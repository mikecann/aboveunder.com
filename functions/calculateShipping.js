export function handler(event, context, callback) {
    console.log("Calculating shipping", { event: event, context: context });
    callback(null, {
        statusCode: 200,
        body: JSON.stringify({ msg: "Calculated!!" })
    });
}
