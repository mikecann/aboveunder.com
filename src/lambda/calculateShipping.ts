import { LambdaEvent, LambdaCallback } from "./types";

const exampleBody = {
  "eventName": "shippingrates.fetch",
  "mode": "Live",
  "createdOn": "2015-02-21T14:58:02.6738454Z",
  "content": {
    "token": "22808196-0eff-4a6e-b136-3e4d628b3cf5",
    "creationDate": "2015-02-21T14:58:02.6738454Z",
    "modificationDate": "2015-02-21T14:58:02.6738454Z",
    "status": "Processed",
    "currency": "USD",
    "paymentMethod": "CreditCard",
    "email": "customer@snipcart.com",
    "cardHolderName": "Nicolas Cage",
    "billingAddressName": "Nicolas Cage",
    "billingAddressCompanyName": "Company name",
    "billingAddressAddress1": "888 The street",
    "billingAddressAddress2": "",
    "billingAddressCity": "QuÃ©bec",
    "billingAddressCountry": "CA",
    "billingAddressProvince": "QC",
    "billingAddressPostalCode": "G1G 1G1",
    "billingAddressPhone": "(888) 888-8888",
    "shippingAddressName": "Nicolas Cage",
    "shippingAddressCompanyName": "Company name",
    "shippingAddressAddress1": "888 The street",
    "shippingAddressAddress2": "",
    "shippingAddressCity": "QuÃ©bec",
    "shippingAddressCountry": "CA",
    "shippingAddressProvince": "QC",
    "shippingAddressPostalCode": "G1G 1G1",
    "shippingAddressPhone": "(888) 888-8888",
    "shippingAddressSameAsBilling": true,
    "finalGrandTotal": 310.00,
    "shippingAddressComplete": true,
    "creditCardLast4Digits": "4242",
    "shippingFees": 10.00,
    "shippingMethod": "Livraison",
    "items": [{
      "uniqueId": "eb4c9dae-e725-4dad-b7ae-a5e48097c831",
      "token": "22808196-0eff-4a6e-b136-3e4d628b3cf5",
      "id": "1",
      "name": "Movie",
      "price": 300.00,
      "originalPrice": 300.00,
      "quantity": 1,
      "url": "https://snipcart.com",
      "weight": 10.00,
      "description": "Something",
      "image": "http://placecage.com/50/50",
      "customFieldsJson": "[]",
      "stackable": true,
      "maxQuantity": null,
      "totalPrice": 300.0000,
      "totalWeight": 10.00
    }],
    "subtotal": 610.0000,
    "totalWeight": 20.00,
    "discounts": [],
    "willBePaidLater": false
  }
}

type HTTPBody = typeof exampleBody;

const exampleResponse = {
  "rates": [{
    "cost": 10,
    "description": "10$ shipping"
    }, {
    "cost": 20,
    "description": "20$ shipping",
    "guaranteedDaysToDelivery": 5
    },
  ]
}

type SuccessResponse = typeof exampleResponse;

// const exampleError = {
//   "errors": [{
//     "key": "invalid_postal_code",
//     "message": "The postal code is invalid."
//     },
//   ]
// };

// type ErrorResponse = typeof exampleError;

export function handler(event: LambdaEvent<HTTPBody>, context:any, callback: LambdaCallback) {

    const response : SuccessResponse = {
      rates: [
        { cost: 33, description: "33 bucks shipping" }
      ]
    }

    callback(null, {
      statusCode: 200,
      body: JSON.stringify(response)
    })

  }