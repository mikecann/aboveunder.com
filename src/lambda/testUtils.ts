import { LambdaFunction, LambdaCallbackResult, LambdaCallResult, LambdaCallbackError } from "./types";

export async function callFunction(event:any, context:any, lambda: LambdaFunction) : Promise<LambdaCallbackResult>
{
    return new Promise<LambdaCallbackResult>(resolve => {
        lambda(event, context, (error?:LambdaCallbackError, result?:LambdaCallResult) => {
            const callbackResult : LambdaCallbackResult = { error: error || null, result: result || null };
            resolve(callbackResult);
        })
    });
}