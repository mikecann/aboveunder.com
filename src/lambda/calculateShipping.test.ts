import { handler, HTTPBody } from "./calculateShipping";
import { callFunction } from "./testUtils";
import { LambdaEvent } from "./types";

test('no event returns error', async () => 
{
    const result = await callFunction({},{}, handler);
    expect(result).toMatchSnapshot();
});

test('GET event returns error', async () => 
{
    const event : LambdaEvent = {
      httpMethod: "GET",
      body: "",
      headers: {},
      queryStringParameters: ""
    }
    const result = await callFunction(event,{}, handler);
    expect(result).toMatchSnapshot();
});

test('it returns the correct response', async () => 
{
    const event : LambdaEvent = {
      httpMethod: "POST",
      body: JSON.stringify({
        content: {
          items: [{
            
          }]
        }
      }),
      headers: {},
      queryStringParameters: ""
    }

    const result = await callFunction(event, {}, handler);
    expect(result.result.body).toMatch(JSON.stringify({
      rates: [
        { cost: 33, description: "33 bucks shipping" }
      ]
    }))
});

