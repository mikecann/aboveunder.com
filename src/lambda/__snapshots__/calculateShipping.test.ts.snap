// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET event returns error 1`] = `
Object {
  "error": null,
  "result": Object {
    "body": "{\\"errors\\":[{\\"key\\":\\"exception\\",\\"message\\":\\"Error: Improperly formatted event\\"}]}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "statusCode": 500,
  },
}
`;

exports[`no event returns error 1`] = `
Object {
  "error": null,
  "result": Object {
    "body": "{\\"errors\\":[{\\"key\\":\\"exception\\",\\"message\\":\\"Error: Improperly formatted event\\"}]}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "statusCode": 500,
  },
}
`;
